otter/log/cloudfeeds.py

```diff
@@ -0,0 +1,177 @@
+from otter.util.pure_http import has_code
+from otter.util.retry import (
+    exponential_backoff_interval,
+    retry_effect,
+    retry_times)
+
+
+class UnsuitableMessage(Exception):
+    """
+    Raised when message is not suitable to be pushed to cloud feed
+    """
+    def __init__(self, message):
+        self.unsuitable_message = message
+
+
+# Mapping of items in a log event to cloud feeds event
+log_cf_mapping = {
+    "scaling_group_id": "scalingGroupId",
+    "policy_id": "policyId",
+    "webhook_id": "webhookId",
+    "username": "username",
+    "desired_capacity": "desiredCapacity",
+    "current_capacity": "currentCapacity",
+    "message": "message"
+}
+
+
+def sanitize_event(event):
+    """
+    Sanitize event by removing all items except the ones in autoscale schema.
+
+    :param dict event: Event to sanitize as given by Twisted
+
+    :return: (dict, bool, str) tuple where dict -> sanitized event,
+        bool -> is it error event?, str -> ISO8601 formatted UTC time of event
+    """
+    cf_event = {}
+    error = False
+
+    # format message
+    event_copy = deepcopy(event)
+    PEP3101FormattingWrapper(lambda e: None)(event_copy)
+    msg = event_copy["message"]
+    event["message"] = msg[0]
+
+    # map keys in event to CF keys
+    for log_key, cf_key in log_cf_mapping.iteritems():
+        if log_key in event:
+            cf_event[cf_key] = event[log_key]
+
+    if event.get('isError', False):
+        error = True
+        if ('traceback' in cf_event['message'] or
+           'exception' in cf_event['message']):
+            raise UnsuitableMessage(cf_event['message'])
+
+    return (cf_event, error,
+            datetime.utcfromtimestamp(event["time"]).isoformat())
+
+
+request_format = {
+    "entry": {
+        "@type": "http://www.w3.org/2005/Atom",
+        "title": "autoscale",
+        "content": {
+            "event": {
+                "@type": "http://docs.rackspace.com/core/event",
+                "id": "",
+                "version": "2",
+                "eventTime": "",
+                "type": "INFO",
+                "region": "",
+                "product": {
+                    "@type": "http://docs.rackspace.com/event/autoscale",
+                    "serviceCode": "Autoscale",
+                    "version": "1",
+                    "message": ""
+                }
+            }
+        }
+    }
+}
+
+
+def prepare_request(req_fmt, event, error, timestamp, region):
+    """
+    Prepare request based on request format
+
+    :returns: Effect of `str` containing request
+    """
+    request = deepcopy(req_fmt)
+    if error:
+        request['entry']['content']['event']['type'] = 'ERROR'
+    request['entry']['content']['event']['region'] = region
+    request['entry']['content']['event']['eventTime'] = timestamp
+    request['entry']['content']['event']['product'].update(event)
+
+    def set_id(_id):
+        request['entry']['content']['event']['id'] = _id
```

Comment:
This isn't quite what I imagined. While technically prepare_request, as a white box, is pure, I think it can be factored a bit more elgeantly without this mutation here.

What I suggest is this: make prepare_request take a parameter for the event ID. In add_event, create an Effect of Func(uuid4), and then attach prepare_request as a callback to it, which will get the event ID (and all the other parameters it currently takes) and return a new request dict. This makes prepare_request a simple function that doesn't care about effects and just produces a pure data structure.